### PR TITLE
WooCommerce: Add state for product listing

### PR DESCRIPTION
**Pull request by justinshreve**
This PR adds a fetchProducts action, and related reducers and selectors for fetching a list of products.

Quick note on PR size: The actual logic/meat of this PR is only about ~180 lines. The rest is tests/test related code, and documentation.

This is patterned off of #14831 so the pagination system will work similarly for products and orders.  cc @ryelle as this also implements a way to get the total pages from the WooCommerce API.

To Test:
* `npm test` and make sure all tests pass (`npm run test-client client/extensions/woocommerce/state/sites/products` to run just products tests)

### CODE REVIEW

File: client/extensions/woocommerce/state/sites/products/selectors.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	return false;
+};
+
+/**
+ * @param {Object} state Whole Redux state tree
+ * @param {Number} [siteId] Site ID to check. If not provided, the Site ID selected in the UI will be used
+ * @return {Number} Total number of pages of products available on a site, or 1 if not loaded yet.
+ */
+export const getTotalProductsPages = ( state, siteId = getSelectedSiteId( state ) ) => {
+	return get( state, [ 'extensions', 'woocommerce', 'sites', siteId, 'products', 'totalPages' ], 1 );
```
Comment (allendav) [inline client/extensions/woocommerce/state/sites/products/selectors.js]: If we have no idea how many product pages there are, wouldn't `0` make more sense as a default?

**Comment by allendav** [comment]
Probably need to define the number of products to fetch in a page on calypso side - we don't want to let the client drive this - they could filter in a ridiculous value and cause performance problems.

**Comment by allendav** [comment]
As we chatted about in slack, if we have lower numbered pages contain newer products, we're going to have all our pages invalidated everytime we add a new product.  Please consider having lower numbered pages contain older products. That way when new products are added, the number of "pages" that are invalidated are limited to 1 or 2.

That does admittedly bring up another issue in that storage pages might contain fewer products than a UI page would want, but I think we could get around that by having the UI load the "last" product of products initially and then, if that is less than the UI wants, also then append some set of products from the next storage page.

**Comment by justinshreve** [comment]
> Probably need to define the number of products to fetch in a page on calypso side - we don't want to let the client drive this - they could filter in a ridiculous value and cause performance problems.

Good call. Handled in https://github.com/Automattic/wp-calypso/pull/15065/commits/48e960c34a76685ba071030d9bf6009cd3cbdbbe

**Comment by justinshreve** [comment]
Okay I've updated this based on the slack convo and what we are doing for v1. I've removed pages from state. I've kept the total amount of pages on the server, and I've kept `isLoading` as is, so we can still keep track of separate fetch calls that might be running.

**Pull request merged**